### PR TITLE
Bunch of fixes and improvements

### PR DESCRIPTION
**Pull request by kurotatsu77**
Added replacement for standard dart traps, turning them into black goo traps, added UD_TrapManifest_ME_Script. NPCs are supported. 

Made UnforgivingSkyrim.esp patch for easier changes merge, to address GitHub inability to show changes inside esp. Suggest all contributors to create their own delta esps and commit them.

Fixed Abadon Cursed StraitJacket script item keyword from armbinder to straitjacket. 
Corrected name of Purified Goo magic effect

Changed Abadon Chastity Bra and Belt to looks shiny like cursed items from DCUR

Expanded FormID lists to include more hard bondage items

Wrote some massive changes into manifest random restraint procedure to make it use DD framework leveled lists and added harnesses and corsets. As result filters were changed massively, it was fixed to work again, but bit coding is changed. Abadon weapon and plug scripts were fixed to use proper filter values.

Changed devices in Abadon simple outfit - swapped iron boots for nicer looking black leather ones. 
Changed final punishing Abadon outfit, added more devices to be worn to ensure proper punishment.

Corrected some spelling and grammatical errors in messages.

### CODE REVIEW

**Comment by IHateMyKite** [comment]
Nice work, will test it once I have some free time. The idea with .esp patches for github is also great. I think I will add Code of conduct so there is some unification between all the changes. For now, I would only ask you to rename UnforgivingSkyrim.esp so it also contains the number of pull request (so something like this -> PR96_UnforgivingSkyrim.esp) so it will be easier for me to merge it in the future. Once I merge it, I will just remove the patch.

**Comment by kurotatsu77** [comment]
Yes, no problem.
Done.

**Comment by kurotatsu77** [comment]
I think I marked this as ready for review, @IHateMyKite

File: Scripts/Source/UD_PurifiedGoo_AME.psc [hunk 1]
```diff
@@ -6,9 +6,9 @@ UDCustomDeviceMain Property UDCDmain auto
 Event OnEffectStart(Actor akTarget, Actor akCaster)
     UDCDmain.DisableActor(akTarget)
     int loc_filter = 0xffffffff
-    if Utility.randomInt(0,1) ;50 % chance for suit
-        loc_filter = Math.LogicalAnd(loc_filter,0xFFFFFDFF)
-    endif
+;    if Utility.randomInt(0,1) ;50 % chance for suit
```
Comment (IHateMyKite) [inline Scripts/Source/UD_PurifiedGoo_AME.psc:9]: I think that previous implementation was better, as it allowed the player also see the belt/bra sometimes. Without this, the catsuit will always be equipped.

File: Scripts/Source/UD_RandomRestraintManager.psc [hunk 2]
```diff
@@ -80,32 +80,43 @@ bool _mutex = false
...
-    UD_CheckKeywords.addForm(libs.zad_DeviousSuit)            ;iIndex = 9
-    UD_CheckKeywords.addForm(libs.zad_DeviousPlugVaginal)    ;iIndex = 10
-    UD_CheckKeywords.addForm(libs.zad_DeviousPlugAnal)        ;iIndex = 11
-    
-    UD_CheckKeywords.addForm(libs.zad_DeviousPiercingsVaginal)    ;iIndex = 12
-    UD_CheckKeywords.addForm(libs.zad_DeviousPiercingsNipple)    ;iIndex = 13
-    UD_CheckKeywords.addForm(libs.zad_DeviousGloves)            ;iIndex = 14
-    UD_CheckKeywords.addForm(libs.zad_deviousHood)                 ;iIndex = 15
-    
+
```
Comment (IHateMyKite) [inline Scripts/Source/UD_RandomRestraintManager.psc:94]: Rearranging the keywords will break previously created saves for older UD version. My current policy is to make new version UD backward compatible. If you still want to implement it like this, you will have to add and hook Update function, so when the game is loaded, the array is wiped and filled with rearranged values. In that case, also, please update the documentation which is at top of the script

File: Scripts/Source/UD_AbadonWeapon_AME.psc [hunk 3]
```diff
@@ -18,7 +18,7 @@ Function OnEffectStart(Actor akTarget, Actor akCaster)
         if UDmain.TraceAllowed()
             UDmain.Log(GetActorName(akTarget) + " hit by abadon weapon. Locking device")
         endif
-        UDmain.UDRRM.LockRandomRestrain(loc_actor,false,0x00000080) ;lock hand restrain
+        UDmain.UDRRM.LockRandomRestrain(loc_actor,false,0x00000010) ;lock hand restraint
```
Comment (IHateMyKite) [inline Scripts/Source/UD_AbadonWeapon_AME.psc:21]: Either update the UDRRM or undo the changes (see comment in UDRRM)

File: Scripts/Source/UD_AbadonPlug_RenderScript.psc [hunk 4]
```diff
@@ -188,18 +188,18 @@ Function randomEquipHandRestrain()
...
         else
-            UDCDmain.Print(getDeviceName() + " locks restrain on " + getWearerName())
+            UDCDmain.Print(getDeviceName() + " locks restraint on " + getWearerName())
         endif
     endif
 EndFunction
 
 bool Function equipRandomRestrain()
-    Armor device = UDmain.UDRRM.getRandomSuitableRestrain(getWearer(),0xFBFB)
+    Armor device = UDmain.UDRRM.getRandomSuitableRestrain(getWearer(),0xFFFCFFF3) ; exclude belts and plugs
```
Comment (IHateMyKite) [inline Scripts/Source/UD_AbadonPlug_RenderScript.psc:197]: Either update the UDRRM or undo the changes (see comment in UDRRM)

File: Scripts/Source/UD_TrapManifest_ME_Script.psc [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    if !akTarget.wornhaskeyword(libs.zad_deviousheavybondage)
+        UDCDmain.DisableActor(akTarget)
+
+		bool loc_haveSuit = akTarget.wornhaskeyword(libs.zad_deviousSuit)
+		Formlist loc_formlist = none
+	    Armor device = none
+        Int loc_arousal = UDCDmain.UDOM.getArousal(akTarget)
+
+;        if Utility.randomInt(1,99) < Round(UDCDMain.UD_BlackGooRareDeviceChance*fRange(loc_arousal/50,1.0,2.0))
```
Comment (IHateMyKite) [inline Scripts/Source/UD_TrapManifest_ME_Script.psc:22]: Remove long comment block

File: Scripts/Source/UD_TrapManifest_ME_Script.psc [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        
+		if random > 90
+		    loc_formlist = UDCDmain.UDmain.UDRRM.UD_AbadonDeviceList_HeavyBondageHard
+        elseif random > 60 
+			loc_formlist = UDCDmain.UDmain.UDRRM.UD_AbadonDeviceList_HeavyBondage
+		else
+		    loc_formlist = UDCDmain.UDmain.UDRRM.UD_AbadonDeviceList_HeavyBondageWeak
+        endif
+
+        if loc_haveSuit
```
Comment (IHateMyKite) [inline Scripts/Source/UD_TrapManifest_ME_Script.psc:74]: Rearrange the lines code, so the script is more readable

**Review by IHateMyKite — CHANGES_REQUESTED** [review]
Please update the PR96_UnforgivingSkyrim.esp . The .esp is using the Cursed Loot textures. This means that the Cursed Loot is now UD master, which should not be the case. Either use different texture or undo the changes, so the Cursed Loot is no master.

File: Scripts/Source/UD_PurifiedGoo_AME.psc (hunk 1 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_PurifiedGoo_AME.psc:9]: Actually, this random is redundant, thanks to the keywords formlist rearranged, thus ensuring that HeavyBondage it applied first, and suit is after, only if that keyword available. All random calls are already inside RandomManifest.

File: Scripts/Source/UD_RandomRestraintManager.psc (hunk 2 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_RandomRestraintManager.psc:94]: Well, it has to be re-filled anyways because of added 2 keywords for corsets and harnesses.

I'm not sure how to add that Update function, any pointers for me?

Yep, will update the documentation so it will state clearly what is what. Makes total sense. I've left it in this state to be able to see "before" and "after" and assess changes.

File: Scripts/Source/UD_AbadonWeapon_AME.psc (hunk 3 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_AbadonWeapon_AME.psc:21]: Of course.

File: Scripts/Source/UD_AbadonPlug_RenderScript.psc (hunk 4 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_AbadonPlug_RenderScript.psc:197]: Yep.

File: Scripts/Source/UD_TrapManifest_ME_Script.psc (hunk 5 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_TrapManifest_ME_Script.psc:22]: Ah, my blasted habit of avoiding removing old code and preserve it in comments...
Will clean it up.

**Comment by IHateMyKite** [comment]
Also, one more point, please don't use tabs. They work badly on github. Use spaces instead. If you use Notepad++, you can just edit settings, so tabs are replaces by spaces (in case of UD by 4 spaces). You can also use Notepad++ to replace currently places tabs by replacing all \t characters with 4 spaces "    ".

**Comment by IHateMyKite** [comment]
It is never a good idea to use tabs for programming, as different text editors will use different spacing for tabs. Because of that the text which allign fine for you will not allign good for other people (like me in this case because of how github handles tabs). By using spaces, you can never go wrong as the text will always be aligned correctly, no matter what text editor you use.

But in case of writing texts in Word, it is perfect habit as it saves you time.

File: Scripts/Source/UD_PurifiedGoo_AME.psc (hunk 1 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_PurifiedGoo_AME.psc:9]: > Please update the PR96_UnforgivingSkyrim.esp . The .esp is using the Cursed Loot textures. This means that the Cursed Loot is now UD master, which should not be the case. Either use different texture or undo the changes, so the Cursed Loot is no master.

Will copying TXTS from Cursed Loot, still pointing to the Cursed Loot files and removed master will be OK?
Or we definitely need to make copies of those files into UD assets?

Comment (kurotatsu77) [reply Scripts/Source/UD_PurifiedGoo_AME.psc]: OK, esp has no CursedLott as master anymore, made copies of TXTS and ARMA records, still using files from DCL. Is this OK?
Reverted change in the PurifiedGoo script, filter is set properly.

File: Scripts/Source/UD_RandomRestraintManager.psc (hunk 2 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_RandomRestraintManager.psc:99]: I think I figured out Update() function, added hook into main UD script, added necessary functions in UDRRM, tested in old save - it worked.

Cleaned up documentation, it's all written now as it should.

File: Scripts/Source/UD_AbadonPlug_RenderScript.psc (hunk 4 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_AbadonPlug_RenderScript.psc:197]: Done.

File: Scripts/Source/UD_AbadonWeapon_AME.psc (hunk 3 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_AbadonWeapon_AME.psc:21]: Done.
Tested, working.

File: Scripts/Source/UD_TrapManifest_ME_Script.psc (hunk 5 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_TrapManifest_ME_Script.psc]: Cleaned it up, should be good now.

**Comment by kurotatsu77** [comment]
Went through all changed scripts, replaced tabs with spaces.

File: Scripts/Source/UD_RandomRestraintManager.psc [hunk 7]
```diff
@@ -275,7 +285,7 @@ bool Function LockAnyRandomRestrain(Actor akActor,int iNumber = 1,bool bForce =
     while iNumber
         iNumber -= 1
         loc_res = loc_res || LockRandomRestrain(akActor)
-		Utility.wait(3)	 ; small delay to ensure of no equips running into one another
+        Utility.wait(1)     ; small delay to ensure of no equips running into one another
```
Comment (IHateMyKite) [inline Scripts/Source/UD_RandomRestraintManager.psc:288]: This should be not needed, as lock function is mutexed. Adding this will make locking the devices slower with not additional safety added.

File: Scripts/Source/UD_RandomRestraintManager.psc [hunk 8]
```diff
@@ -342,7 +352,7 @@ int Function LockAllSuitableRestrains(Actor akActor,Bool force = false,Int iPref
             endif
         endif
         loc_i += 1
-		Utility.wait(3)	 ; small delay to ensure of no equips running into one another
+        Utility.wait(3)     ; small delay to ensure of no equips running into one another
```
Comment (IHateMyKite) [inline Scripts/Source/UD_RandomRestraintManager.psc:355]: Same as previous comment

**Review by IHateMyKite — CHANGES_REQUESTED** [review]
The issue with textures can't be solved by pointing at the textures. That will solve the maser issue, but the retextured devices will lack textures if user have no DCUR installed. I think that better way of doing this will be to hook the texture set to previous used texture (or other one, but it must be in base DD). Then, the DCUR patch can be updated so it also edits the texture sets. This way, if user will have DCUS patch installed, the DCUR textures will be used, Otherwise, previous textures will be used.

**Comment by kurotatsu77** [comment]
Done, as bonus I've made MCM page for user-friendly device filter selection.

File: Scripts/Source/UD_RandomRestraintManager.psc [hunk 9]
```diff
@@ -51,68 +53,85 @@ Function StartMutex()
...
     endif
     ready = True
 EndEvent
 
+Function Update()
+    FillOutCheckKeywords()
+    if UDmain.TraceAllowed()    
+        UDCDmain.Log("Refilled Keywords formlist")
+    endif
+    UD_RandomDevice_GlobalFilter = 0xFFFFFFFF
```
Comment (IHateMyKite) [inline Scripts/Source/UD_RandomRestraintManager.psc:105]: This line should not exist. This will wipe all MCM settings which you added on every game reload.

File: Scripts/Source/UD_RandomRestraintManager.psc [hunk 10]
```diff
@@ -162,51 +181,87 @@ Armor Function getRandomDeviceByKeyword(Actor akActor,Keyword kwKeyword)
...
+    elseif kwKeyword == libs.zad_DeviousCorset && !akActor.wornhaskeyword(libs.zad_DeviousHarness)
+        LL = zadDL.zad_dev_corsets
+    elseif kwKeyword == libs.zad_DeviousHarness && !akActor.wornhaskeyword(libs.zad_DeviousCorset)
+        LL = zadDL.zad_dev_harnesses
+    else
+        return none
+    endif
+
+    if LL
+        int tries = 10                        ; 10 attempts
```
Comment (IHateMyKite) [inline Scripts/Source/UD_RandomRestraintManager.psc:252]: Reduce this number to ~2. If it failed first time, there is big chance it will fail again on next attempt. So forcing this check 10 times can freeze player for few minutes before all conflicts are resolved. You should also add some console trace, so users can see which devices are conflicting. In this case, I think the best option is UDmain.Warning or just GWarning

**Review by IHateMyKite — CHANGES_REQUESTED** [review]
Both of following issues are in esp, so you can skip them and I will fix them when I will be merging them. Or you can do it yourself, I will keep it on you. Once you fix the 2 script related issues in UD_RandomRestraintManager, I will merge the PR.

[PR96_UnforgivingSkyrim.esp](https://github.com/IHateMyKite/UnforgivingDevices/pull/96/files#diff-b18eaf4593540e767c05dcc2c36647b884c74c2240244eb0fe91e8f3426e0f72)
The formlists should not include Abadon Cursed Straitjacket. Device should be Abadon quest exclusive. Making the device 100% sentient is also crazy, as it will make escaping it almost impossible. 

[PR96_UnforgivingDevices - DCL Patch Patched.esp](https://github.com/IHateMyKite/UnforgivingDevices/pull/96/files#diff-9f1425d667655d79c1639c8d2383500bca82236b896f9b763f9d55f57c963db5)
We didn't understand each other, why I meant was that this esp should only change the texture set, not the device. If it change device, it will mean that every time the base device is changed, the patch device will also need to be updated. Which is pain, and its easy to forget to do it, which results in unnecessary bugs.

File: Scripts/Source/UD_RandomRestraintManager.psc (hunk 9 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_RandomRestraintManager.psc:105]: Oh... *facepalms*

File: Scripts/Source/UD_RandomRestraintManager.psc (hunk 10 above)
Comment (kurotatsu77) [reply Scripts/Source/UD_RandomRestraintManager.psc:252]: Actually, my testing shown that this reroll works really fast!
It's the most useful when pulling various harnesses, some can have collars, some can have chastity.
So it tries several times to pull variety of device which is not in conflict with others.

Maybe leave it as it is for testing, and if some really nasty hold-ups arise, then reduce?

**Comment by kurotatsu77** [comment]
Try it, current Cursed SJ 100% sentient can be escaped easier than Rare Blue Straitjacket.
Also in description there's line about it always activating plugs on escape attempt...
If you feel it's quest-only thing - sure thing, set it as you feel it should be.

File: Scripts/Source/UD_RandomRestraintManager.psc (hunk 10 above)
Comment (IHateMyKite) [reply Scripts/Source/UD_RandomRestraintManager.psc:251]: I have tested it and it took very long time for all devices to be locked. But it might have been caused by something else. Will have to investigate further later. For now, I will ignore this point. After merge, I will have to do some more benchmark testing.

**Review by IHateMyKite — APPROVED** [review]

**Comment by IHateMyKite** [comment]
Nice work. Closing and merging.
